### PR TITLE
Coding interview post

### PR DESCRIPTION
**Pull request by stefan-aws**

### CODE REVIEW

File: assets/src/cracking-the-coding-interview-permutations/cracking-the-coding-interview-in-dafny-permutations.dfy
Comment (alex-chew) [inline assets/src/cracking-the-coding-interview-permutations/cracking-the-coding-interview-in-dafny-permutations.dfy:1]: Does this file verify successfully as written, or does it need modifications (particularly `opaque` and `reveal`)?

**Review by alex-chew — COMMENTED** [review]
The content looks good to me, but I'd want to make sure we only publish source code that verifies as written.

File: _posts/2023-11-08-cracking-the-coding-interview-in-dafny-permutations.markdown [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
...ng Dafny, we were able to i) translate the informal task of writing a function that returns the set of all permutations of a given sequence into a formal, mathematical specification; ii) implement a potential solution to the problem — in this case, in a purely functional style; and iii) formally verify, with a mathematical proof, that the proposed solution is correct. Along the way, we encountered Dafny’s [`seq`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-sequences), [`set`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-sets), [`iset`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-sets), and [`multiset`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-multisets) types, discussed [type parameters](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-type-parameters), proved theorems by induction, reasoned equationally with the [`calc`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-calc-statement) statement, and made heavy use of Dafny’s capability of divide-and-conquer complex lemmas into smaller, feasible components. I hope you enjoyed the ride. The full Dafny source code of this post is available [here](../../../../assets/src/cracking-the-coding-interview-permutations/permutations.dfy).
```
Comment (atomb) [inline _posts/2023-11-08-cracking-the-coding-interview-in-dafny-permutations.markdown:277]:
```suggestion
Using Dafny, we were able to i) translate the informal task of writing a function that returns the set of all permutations of a given sequence into a formal, mathematical specification; ii) implement a potential solution to the problem — in this case, in a purely functional style; and iii) formally verify, with a mathematical proof, that the proposed solution is correct. Along the way, we encountered Dafny’s [`seq`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-sequences), [`set`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-sets), [`iset`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-sets), and [`multiset`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-multisets) types, discussed [type parameters](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-type-parameters), proved theorems by induction, reasoned equationally with the [`calc`](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-calc-statement) statement, and made heavy use of Dafny’s capability to subdivide complex lemmas into smaller, feasible components. I hope you enjoyed the ride. The full Dafny source code of this post is available [here](../../../../assets/src/cracking-the-coding-interview-permutations/permutations.dfy).
```

**Review by atomb — APPROVED** [review]
Looks nice! I commented on a few things that could potentially change, but don't necessarily need to.

I'm also interested in the answer to Alex's question.

File: assets/src/cracking-the-coding-interview-permutations/cracking-the-coding-interview-in-dafny-permutations.dfy
Comment (stefan-aws) [reply assets/src/cracking-the-coding-interview-permutations/cracking-the-coding-interview-in-dafny-permutations.dfy:1]: The file doesn't contain any `opaque` nor `reveal` keywords. (I removed them after your first review.) It verifies as written.

**Review by alex-chew — APPROVED** [review]
LGTM! Nice work :)

**Pull request merged**